vector/src/main/java/im/vector/app/features/onboarding/ftueauth/CaptchaWebview.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2022 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.onboarding.ftueauth
+
+import android.annotation.SuppressLint
+import android.content.DialogInterface
+import android.graphics.Bitmap
+import android.net.http.SslError
+import android.os.Build
+import android.view.KeyEvent
+import android.view.View
+import android.webkit.SslErrorHandler
+import android.webkit.WebResourceRequest
+import android.webkit.WebResourceResponse
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.core.view.isVisible
+import androidx.fragment.app.Fragment
+import com.google.android.material.dialog.MaterialAlertDialogBuilder
+import im.vector.app.R
+import im.vector.app.core.utils.AssetReader
+import im.vector.app.features.login.JavascriptResponse
+import im.vector.app.features.onboarding.OnboardingViewState
+import org.matrix.android.sdk.internal.di.MoshiProvider
+import timber.log.Timber
+import java.net.URLDecoder
+import java.util.Formatter
+import javax.inject.Inject
+
+class CaptchaWebview @Inject constructor(
+        private val assetReader: AssetReader
+) {
+
+    @SuppressLint("SetJavaScriptEnabled")
+    fun setupWebView(
+            container: Fragment,
+            webView: WebView,
+            progressView: View,
+            siteKey: String,
+            state: OnboardingViewState,
+            onSuccess: (String) -> Unit
+    ) {
+        webView.settings.javaScriptEnabled = true
+
+        val reCaptchaPage = assetReader.readAssetFile("reCaptchaPage.html") ?: error("missing asset reCaptchaPage.html")
+
+        val html = Formatter().format(reCaptchaPage, siteKey).toString()
+        val mime = "text/html"
+        val encoding = "utf-8"
+
+        val homeServerUrl = state.selectedHomeserver.upstreamUrl ?: error("missing url of homeserver")
+        webView.loadDataWithBaseURL(homeServerUrl, html, mime, encoding, null)
+        webView.requestLayout()
+
+        webView.webViewClient = object : WebViewClient() {
+            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
+                super.onPageStarted(view, url, favicon)
+
+                if (!container.isAdded) {
+                    return
+                }
```

Comment:
I'm easy here as the scope is so small, happy to do the change. I definitely prefer single returns per method

Reply:
72740b5dea6ddb104c09fddbcb97e36356976d22

Reply:
`single returns per method` means avoid using `return`, no? 😄

Reply:
I would argue for no returns being equal to 1 return, but implicitly on the last line :sweat_smile: the main case would be to avoid 

```kotlin
if (foo) return null
if (bar) return null

// and instead combing the cases into...

return when {
  foo -> null
  bar -> null
  else -> // some logic
}
```

Reply:
Oh yes of course for `fun`s which actually return something :). I was thinking about `method` :)